src/DataSets.jl

```diff
@@ -188,21 +201,86 @@ To open a directory as a browsable tree object,
 open(BlobTree, dataset("a_tree_example"))
 ```
 """
-function dataset(proj::AbstractDataProject, name::AbstractString)
-    # Non-fancy search... for now :)
-    # In the future, we can consider parsing `name` into a dataset prefix and a
-    # data selector / resource section. Eg a path for BlobTree which gives us a
-    # SubDataSet
-    #
-    # The URN RFC8141 has some good design inspiration here, in particular the
-    # distinction between r-component and q-component seems relevant:
-    # * Some parameters may need to be passed to the "resolver" (ie, the data
-    #   storage backend)
-    # * Some parameters may need to be passed to the dataset itself (eg, a
-    #   relative path within the dataset)
+function dataset(proj::AbstractDataProject, spec::AbstractString)
+    namestr, querystr, fragmentstr = _split_dataspec(spec)
+
+    if isnothing(namestr)
+        throw(ArgumentError("Invalid dataset specification: $spec"))
+    end
+
+    dataset = proj[namestr]
+
+    if isnothing(querystr) && isnothing(fragmentstr)
+        return dataset
+    end
+
+    # Enhance dataset with "parameters" holding URL-like fragment & query
+    # params
     #
-    # See https://datatracker.ietf.org/doc/html/rfc8141#page-12
-    return proj[name]
+    # We need to take care here to add copy()s to avoid modifying the original
+    # dataset configuration.
+    conf = copy(dataset.conf)
+    params = haskey(conf, "parameters") ? copy(conf["parameters"]) : Dict()
+    conf["parameters"] = params
+    if !isnothing(querystr)
+        query = Dict(split.(split(querystr, '&'), '='))
+        params["query"] = haskey(params, "query") ?
+            merge(query, params["query"]) : query
+    end
+    if !isnothing(fragmentstr)
+        params["fragment"] = fragmentstr
+    end
+
+    return DataSet(conf)
+end
+
+
+# Percent-decode a string according to the URI escaping rules.
+# Vendored from URIs.jl for now to avoid depending on that entire package for
+# this one function.
+function _unescapeuri(str)
+    occursin("%", str) || return str
+    out = IOBuffer()
+    i = 1
+    io = IOBuffer(str)
+    while !eof(io)
+        c = read(io, Char)
+        if c == '%'
+            c1 = read(io, Char)
+            c = read(io, Char)
+            write(out, parse(UInt8, string(c1, c); base=16))
```

Comment:
True! As the comment says I just stole this function from URIs.jl though! So I can't take real credit for any neatness (or bugs :) ) here.

The other way to do this would be to iterate the input string using `i = nextind(str, i)` and `str[i]`. This is also unicode-safe and avoids the need to create an `IOBuffer` just to read the input.

The other-other way to do it would be to iterate the input string using `Iterators.Stateful(str)`, which can be used statefully, much like the `IOBuffer` is here.

Reply:
Oh yes, forget those comments about iterating input — agreed that the output `IOBuffer` is very natural.